### DISCUSSION TITLE
some questions regarding usage inside an editor

### DISCUSSION THREAD

**Discussion by jrekier** [zk-org/zk#37] [Help]
These might be naive questions.

I am unsure of the best way to use `zk` inside a given editor (e.g. vim). When using [neuron](https://neuron.zettel.page/) with its [editor integration](https://neuron.zettel.page/editor) options, there are ways to 'expand' the name of a link (its 'id') to show the note's title. I find this option crucial to edit and link together notes adequately when using random id's. 
How would you do that using `zk`?

Similarly, what would be the adequate way to insert a link inside an existing note currently being edited ? In previous instalments of [neuron.vim](https://github.com/fiatjaf/neuron.vim), you would simply type a command like `gzi`, which would produce an `fzf` window from which you can select the note you wish to insert. 

I am asking because the above mentioned .vim plugins are now partly deprecated or at least they aren't working for me anymore and I wish to switch to `zk` which appears both more powerful and simpler than other options. 

Thanks for the help.

**Comment by mickael-menu** [comment]
Welcome @jrekier!

### Universal (but basic) integration with LSP

`zk` itself [ships with an LSP server](https://microsoft.github.io/language-server-protocol/overviews/lsp/overview/), which can provide basic integration in any LSP-compatible editor. LSP can be used only to implement "simple" features such as auto-completion of tags and links, creation of new notes from a text selection or following (wiki-)links, etc. [See this issue for a comprehensive list of LSP features to come](https://github.com/mickael-menu/zk/issues/22).

However, it can't provide any custom interface such as the `fzf` window. Personally I found that the LSP integration is enough for most of my needs. [You can find instructions to setup LSP in a few editors here](https://github.com/mickael-menu/zk/blob/main/docs/editors-integration.md).

### Advanced integration with editor-specific plugins

For more complex use cases, such as embedding an `fzf` window to create a link, we need to implement editor-specific plugins. `zk` is still young so there's not a lot of solutions out there, but @megalithic is currently working on [a plugin for Neovim 0.5](https://github.com/megalithic/zk.nvim).

### Specific use cases

> When using [neuron](https://neuron.zettel.page/) with its [editor integration](https://neuron.zettel.page/editor) options, there are ways to 'expand' the name of a link (its 'id') to show the note's title.

You can't do that with `zk` at the moment, but I believe it could be implemented with LSP. I still need to figure out the best way to do that but there is an experimental [LSP server for neuron](https://github.com/aca/neuron-language-server) which did something similar. Does this screenshot looks like what you had in mind?

<img src="https://github.com/aca/neuron-language-server/raw/master/images/diagnostic.png" width="300px"/>

> Similarly, what would be the adequate way to insert a link inside an existing note currently being edited ?

With LSP, `zk` can auto-complete links when entering `[[`. This only works well if you know the title or the file path of the note to link, because you can't search in the content of the note in this context. [The format of the inserted link can be customized](https://github.com/mickael-menu/zk/blob/main/docs/note-format.md), by default it's a regular Markdown.

For searching in the content of the notes, we will need plugin-specific support, and I think @megalithic planned to add this to [`zk.nvim`](https://github.com/megalithic/zk.nvim) eventually.

Here's how the auto-completion looks like with LSP:

https://user-images.githubusercontent.com/58686775/116608753-dc564680-a933-11eb-9977-116022ac27f9.mov

Let me know if there are other missing features for you with the editor integration, so I know what to prioritize. I'm currently working on creating new notes from the current selection.

**Reply by jrekier** [reply]
Hi @mickael-menu and thank you very much for your detailed reply.

I'm far from being an expert in LSP, or anything code-related for that matter ... (pretty much the enthusiast autodidact). LSP look very cool ! Looking forward to what might come of it for `zk` in the future. 

Your answers are spot on what I was trying to ask about. What I get from them is that I should probably be on the lookout for @megalithic 's plug-in support updates to do what I have in mind. 😄 

Although, since you mention that the current LSP features are sufficient for your purpose, I would bevery interested to know more about your workflow using `zk`. I think I read somewhere that you use it to maintain a *zettelkasten* type of notebook, which is what I am trying to do. In particular (as my original question indicates), there must be something that I am missing about the right way to link between notes in case the id of files are not explicit (e.g. randomly generated). Autocomplete based on filenames doesn't really do the trick here... 
The feature based on `fzf` in `neuron.vim` used to work wonders for me but has become obsolete when neuron moved on to `2.0`. So much for future proof. 😢 

Thanks for your help!

**Reply by mickael-menu** [reply]
> I'm far from being an expert in LSP, or anything code-related for that matter ... (pretty much the enthusiast autodidact). LSP look very cool !

I've been exploring LSP for only a few weeks myself, but the possibilities are really promising!

> there must be something that I am missing about the right way to link between notes in case the id of files are not explicit (e.g. randomly generated). Autocomplete based on filenames doesn't really do the trick here...

I'm using 4 random chars for my filenames myself, so I think what you want to do is already possible with LSP. In the screencast I shared above, the completion list is actually displaying the notes' titles (taken from the first Markdown heading in the note, or the `title` YAML frontmatter property). And I can filter by either the note title or the file path, which is hidden from the list.

The only thing that fzf would bring to the table would be to search in the notes' content instead of only the title and path.

Note that the auto-completion with Neovim's built-in LSP client is still a bit rough, how to configure it properly is not yet clear to me. But until we have this figured out, [it's working really well with coc.nvim](https://github.com/mickael-menu/zk/blob/main/docs/editors-integration.md#vim-and-neovim-04) for me

I suggest you subscribe to the [dedicated LSP issue](https://github.com/mickael-menu/zk/issues/22) where I'll post any update related to this.

**Reply by jrekier** [reply]
> I'm using 4 random chars for my filenames myself, so I think what you want to do is already possible with LSP. In the screencast I shared above, the completion list is actually displaying the notes' titles (taken from the first Markdown heading in the note, or the `title` YAML frontmatter property). And I can filter by either the note title or the file path, which is hidden from the list.

I had not realized that the drop-down menu listed the titles and not just the id's ! This will already help a lot. Thanks !

> Note that the auto-completion with Neovim's built-in LSP client is still a bit rough, how to configure it properly is not yet clear to me. But until we have this figured out, [it's working really well with coc.nvim](https://github.com/mickael-menu/zk/blob/main/docs/editors-integration.md#vim-and-neovim-04) for me
> 
> I suggest you subscribe to the [dedicated LSP issue](https://github.com/mickael-menu/zk/issues/22) where I'll post any update related to this.

I will also look at both of these. Thank you again. :)

**Reply by mickael-menu** [reply]
@jrekier I just added diagnostics for dead links and wiki-link titles.

Neovim's 0.5 built-in LSP client displays these diagnostics as virtual text at the end of the line, which is useful for `[[id]]` type of links. However I disabled this diagnostic by default as it's not so useful for other editors, you will need to enable it in [your `zk` config](https://github.com/mickael-menu/zk/blob/main/docs/config.md):

```toml
[lsp.diagnostics]
# Report titles of wiki-links as hints.
wiki-title = "hint"
```

Here's a screenshot after customizing the colors:

<details><summary><tt>~/.config/nvim/init.vim</tt></summary>

```viml
highlight LspDiagnosticsDefaultError ctermfg=red guifg=red
highlight LspDiagnosticsUnderlineError ctermfg=red guifg=red
highlight LspDiagnosticsDefaultHint ctermfg=yellow guifg=yellow
highlight LspDiagnosticsUnderlineHint cterm=none gui=none
```
</details>

<img width="600" alt="Screenshot 2021-05-16 at 21 44 05" src="https://user-images.githubusercontent.com/58686775/118411475-142fee80-b695-11eb-83e7-99b9e90942ac.png">


* [zk-v0.4.0-5-gaa68199-macos-x86_64.zip](https://github.com/mickael-menu/zk/files/6489967/zk-v0.4.0-5-gaa68199-macos-x86_64.zip)
* [zk-v0.4.0-5-gaa68199-linux-i386.tar.gz](https://github.com/mickael-menu/zk/files/6489965/zk-v0.4.0-5-gaa68199-linux-i386.tar.gz)
* [zk-v0.4.0-5-gaa68199-linux-arm64.tar.gz](https://github.com/mickael-menu/zk/files/6489964/zk-v0.4.0-5-gaa68199-linux-arm64.tar.gz)
* [zk-v0.4.0-5-gaa68199-linux-amd64.tar.gz](https://github.com/mickael-menu/zk/files/6489962/zk-v0.4.0-5-gaa68199-linux-amd64.tar.gz)

**Reply by jrekier** [reply]
Hi @mickael-menu. This looks awesome! Will give it a look as quickly as possible. 
Thank you very much for notifying as well.

Cheers
